src/hotspot/share/gc/g1/g1FullGCMarker.inline.hpp

```diff
@@ -162,16 +154,31 @@ inline void G1FullGCMarker::follow_object(oop obj) {
 void G1FullGCMarker::drain_stack() {
   do {
     oop obj;
-    while (pop_object(obj)) {
+    while (_oop_stack.pop_overflow(obj)) {
+      if (!_oop_stack.try_push_to_taskqueue(obj)) {
+        assert(_bitmap->is_marked(obj), "must be marked");
+        follow_object(obj);
+      }
+    }
```

Comment:
What do you think about instead of extracting all this into `drain_stack()` have something like `transfer_from_overflow()` as the first statement in in `pop_object()`? It would either return an object that can't be transferred or push it back to the overflow and just let pop do what it does today. 

The problem I see with that is that `pop_object(...)` would have side effects that you wouldn't really expect, so if doing that maybe we should call it `drain_object_stack(...)` or something like that.

Reply:
Refactored the code a bit, to hopefully something better.

I *think* there are some additional gains to be made by not draining the regular oop task queue fully when trying to move work from its overflow stack to the shared one, however measurements were within a few percent so far. I will file an issue for working on that idea - the current change is already much better than what has been there before.

Reply:
I think this is better and I'm not too worried if this is not optimal. As you say, still a good improvement.